### PR TITLE
Specify roles and groups

### PR DESCRIPTION
**Pull request by louischan-oursky**
ref #3537

### CODE REVIEW

File: docs/specs/roles-groups.md [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+# Changes in the headers of resolver
+
+See [./api-resolver.md#x-authgear-user-roles](./api-resolver.md#x-authgear-user-roles).
+
+# Changes in blocking event hooks
+
+Some blocking event hooks can mutate the `roles` and `groups` of a user.
+See [./hook.md#blocking-event-mutations](./hook.md#blocking-event-mutations).
+
+# Changes in user listing
```
Comment (louischan-oursky) [inline docs/specs/roles-groups.md:458]: Clarify that we will take a compromise first.

**Comment by louischan-oursky** [comment]
I have a suggestion. Rename `roles` to `computed_roles` when roles are computed, and `roles` always stand for non-computed roles. Thoughts? @tung2744 @fungc-io @chpapa 

It is because we want to allow the developer to mutate `roles` and `groups` in blocking hooks. Then we should give them `roles` and `groups` in the event payload. In this context, `roles` is non-computed. So `roles` sometimes is computed and sometimes is not, depending what context the developer finds it. I think this will be confusing.

**Comment by fungc-io** [comment]
In our offline discussion,
we plan to use:
- `effective_roles` to represent all roles, including the directly assigned roles and the derived roles
- And `roles` for those directly assigned roles, which can be edited by hooks

**Comment by louischan-oursky** [comment]
@fungc-io The key appearing in JWT and userinfo will be `https://authgear.com/claims/user/effective_roles` instead of `effective_roles`, to be consistent of other Authgear-defined claims.

**Comment by chpapa** [comment]
@louischan-oursky @fungc-io sorry I am confused, so what's the difference between effective_role and role? Why we need two keys?

**Comment by chpapa** [comment]
If the definition of effective_roles is just a union of roles + roles of the group user in.

I think we should either:

- just have roles in user info endpoint and tokens.
- if there are some reason that's confusing (eg when used with admin api), maybe less confusing if we define each users can only belong to groups, and roles are always computed from groups

Definitely not introduce another terms that need a page to explain in doc again 🙃

**Comment by fungc-io** [comment]
Agree to better keep one `roles` only.

@louischan-oursky can you elaborate more on how the user maybe confused by cannot tell "directly assigned roles" and the "derived roles" apart. Perhaps in actual usecase they don't need to tell them apart

**Comment by louischan-oursky** [comment]
We had implemented JWT access token in [PR #750 in Nov 2020](https://github.com/authgear/authgear-server/pull/750), predating [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068.txt), which was published in Oct 2021.

In Section 2.2.3.1. of RFC 9068, it says that `roles` and `groups` should be referred to [RFC 7643](https://www.rfc-editor.org/rfc/rfc7643.txt). In RFC 7643, `roles` is an array of objects. This is different from our definition.

Therefore, we want to avoid using the name `roles` in JWT access token and User Info.

Ben suggested that we simply use the name `https://authgear.com/claims/user/roles` in JWT access token and User Info, to represent effective roles. In event payload and Admin API, `roles` means roles directly associated with the user.

**Comment by fungc-io** [comment]
thanks all

**Comment by louischan-oursky** [comment]
@Ngkaokis Beware of the spec update.

**Pull request merged**